### PR TITLE
fix(inccommand): restrict cmdpreview undo calls

### LINKED ISSUE #20036: heap-use-after-free when inccommand preview callback undos twice
### Neovim version (nvim -v)

v0.8.0-dev-1003-ga62cb406b

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Arch Linux

### Terminal name/version

kitty 0.25.2

### $TERM environment variable

xterm-kitty

### Installation

build from repo

### How to reproduce the issue

1. Run `nvim --clean`
2. Source the following Lua file:
```lua
vim.api.nvim_create_user_command('Foo', function() end, {nargs = '?', preview = function() vim.cmd.undo() end})
vim.api.nvim_input('ifoo<Esc>yypp:')
```
3. Type `Fo`

### Expected behavior

No crash

### Actual behavior

heap-use-after-free:
```
=================================================================
==1171483==ERROR: AddressSanitizer: heap-use-after-free on address 0x61a0000036d8 at pc 0x55c5f03ab7e5 bp 0x7ffc7c651c50 sp 0x7ffc7c651c48
READ of size 4 at 0x61a0000036d8 thread T0
    #0 0x55c5f03ab7e4 in u_undoredo **/src/nvim/undo.c:2240:28
    #1 0x55c5f03a2752 in u_doit **/src/nvim/undo.c:1861:7
    #2 0x55c5f03a1b82 in u_undo **/src/nvim/undo.c:1758:3
    #3 0x55c5ef56673b in ex_undo **/src/nvim/ex_docmd.c:5877:7
    #4 0x55c5ef527462 in execute_cmd0 **/src/nvim/ex_docmd.c:1590:7
    #5 0x55c5ef52281d in execute_cmd **/src/nvim/ex_docmd.c:1662:3
    #6 0x55c5eed9e3aa in nvim_cmd **/src/nvim/api/command.c:636:3
    #7 0x55c5eed17102 in nlua_api_nvim_cmd **/build/src/nvim/auto/lua_api_c_bindings.generated.c:1472:22
    #8 0x55c5f06dbd45 in lj_BC_FUNCC (**/build/bin/nvim+0x270cd45)
    #9 0x55c5f06c8184 in lua_pcall **/.deps/build/src/luajit/src/lj_api.c:1145:12
    #10 0x55c5ef84fbfd in nlua_pcall **/src/nvim/lua/executor.c:140:16
    #11 0x55c5ef859a54 in nlua_do_ucmd **/src/nvim/lua/executor.c:2152:7
    #12 0x55c5f03d43a0 in do_ucmd **/src/nvim/usercmd.c:1559:12
    #13 0x55c5ef526de6 in execute_cmd0 **/src/nvim/ex_docmd.c:1582:13
    #14 0x55c5ef52281d in execute_cmd **/src/nvim/ex_docmd.c:1662:3
    #15 0x55c5ef5cc57c in cmdpreview_may_show **/src/nvim/ex_getln.c:2338:21
    #16 0x55c5ef5bc255 in command_line_changed **/src/nvim/ex_getln.c:2419:10
    #17 0x55c5ef5cae2a in command_line_handle_key **/src/nvim/ex_getln.c:2012:10
    #18 0x55c5ef5ba0fc in command_line_execute **/src/nvim/ex_getln.c:1299:10
    #19 0x55c5f019c70e in state_enter **/src/nvim/state.c:88:26
    #20 0x55c5ef595046 in command_line_enter **/src/nvim/ex_getln.c:821:3
    #21 0x55c5ef59060c in getcmdline **/src/nvim/ex_getln.c:2478:10
    #22 0x55c5ef599f68 in getexline **/src/nvim/ex_getln.c:2725:18
    #23 0x55c5ef4f33d1 in do_cmdline **/src/nvim/ex_docmd.c:519:30
    #24 0x55c5efb9c8bb in nv_colon **/src/nvim/normal.c:4050:20
    #25 0x55c5efb88c45 in normal_execute **/src/nvim/normal.c:1172:3
    #26 0x55c5f019c70e in state_enter **/src/nvim/state.c:88:26
    #27 0x55c5efb5a3c0 in normal_enter **/src/nvim/normal.c:471:3
    #28 0x55c5ef87eaa6 in main **/src/nvim/main.c:583:3
    #29 0x7f3e3b2852cf  (/usr/lib/libc.so.6+0x232cf) (BuildId: 9c28cfc869012ebbd43cdb0f1eebcd14e1b8bdd8)
    #30 0x7f3e3b285389 in __libc_start_main (/usr/lib/libc.so.6+0x23389) (BuildId: 9c28cfc869012ebbd43cdb0f1eebcd14e1b8bdd8)
    #31 0x55c5eec0dde4 in _start /build/glibc/src/glibc/csu/../sysdeps/x86_64/start.S:115

0x61a0000036d8 is located 88 bytes inside of 1208-byte region [0x61a000003680,0x61a000003b38)
freed by thread T0 here:
    #0 0x55c5eecc1ca2 in __interceptor_free.part.0 asan_malloc_linux.cpp.o
    #1 0x55c5efa65854 in xfree **/src/nvim/memory.c:128:3
    #2 0x55c5f03c0789 in u_freeentries **/src/nvim/undo.c:2920:3
    #3 0x55c5f03914fb in u_freebranch **/src/nvim/undo.c:2887:5
    #4 0x55c5f03a4d36 in u_undo_and_forget **/src/nvim/undo.c:1813:3
    #5 0x55c5ef5d4d5b in cmdpreview_restore_state **/src/nvim/ex_getln.c:2218:12
    #6 0x55c5ef5cc760 in cmdpreview_may_show **/src/nvim/ex_getln.c:2365:3
    #7 0x55c5ef5bc255 in command_line_changed **/src/nvim/ex_getln.c:2419:10
    #8 0x55c5ef5cae2a in command_line_handle_key **/src/nvim/ex_getln.c:2012:10
    #9 0x55c5ef5ba0fc in command_line_execute **/src/nvim/ex_getln.c:1299:10
    #10 0x55c5f019c70e in state_enter **/src/nvim/state.c:88:26
    #11 0x55c5ef595046 in command_line_enter **/src/nvim/ex_getln.c:821:3
    #12 0x55c5ef59060c in getcmdline **/src/nvim/ex_getln.c:2478:10
    #13 0x55c5ef599f68 in getexline **/src/nvim/ex_getln.c:2725:18
    #14 0x55c5ef4f33d1 in do_cmdline **/src/nvim/ex_docmd.c:519:30
    #15 0x55c5efb9c8bb in nv_colon **/src/nvim/normal.c:4050:20
    #16 0x55c5efb88c45 in normal_execute **/src/nvim/normal.c:1172:3
    #17 0x55c5f019c70e in state_enter **/src/nvim/state.c:88:26
    #18 0x55c5efb5a3c0 in normal_enter **/src/nvim/normal.c:471:3
    #19 0x55c5ef87eaa6 in main **/src/nvim/main.c:583:3
    #20 0x7f3e3b2852cf  (/usr/lib/libc.so.6+0x232cf) (BuildId: 9c28cfc869012ebbd43cdb0f1eebcd14e1b8bdd8)

previously allocated by thread T0 here:
    #0 0x55c5eecc2cb9 in malloc (**/build/bin/nvim+0xcf3cb9)
    #1 0x55c5efa65562 in try_malloc **/src/nvim/memory.c:80:15
    #2 0x55c5efa6572c in xmalloc **/src/nvim/memory.c:114:15
    #3 0x55c5f038a658 in u_savecommon **/src/nvim/undo.c:363:13
    #4 0x55c5f038988f in u_save **/src/nvim/undo.c:237:10
    #5 0x55c5efc08f18 in do_put **/src/nvim/ops.c:3072:9
    #6 0x55c5efbcd23b in nv_put_opt **/src/nvim/normal.c:7322:5
    #7 0x55c5efba79f6 in nv_put **/src/nvim/normal.c:7221:3
    #8 0x55c5efb88c45 in normal_execute **/src/nvim/normal.c:1172:3
    #9 0x55c5f019c70e in state_enter **/src/nvim/state.c:88:26
    #10 0x55c5efb5a3c0 in normal_enter **/src/nvim/normal.c:471:3
    #11 0x55c5ef87eaa6 in main **/src/nvim/main.c:583:3
    #12 0x7f3e3b2852cf  (/usr/lib/libc.so.6+0x232cf) (BuildId: 9c28cfc869012ebbd43cdb0f1eebcd14e1b8bdd8)

SUMMARY: AddressSanitizer: heap-use-after-free **/src/nvim/undo.c:2240:28 in u_undoredo
Shadow bytes around the buggy address:
  0x0c347fff8680: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c347fff8690: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c347fff86a0: fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa
  0x0c347fff86b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c347fff86c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c347fff86d0: fd fd fd fd fd fd fd fd fd fd fd[fd]fd fd fd fd
  0x0c347fff86e0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c347fff86f0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c347fff8700: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c347fff8710: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c347fff8720: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1171483==ABORTING
```

### LINKED ISSUE #20248: inccommand: breaking undo chain in insert mode corrupts undo file
### Neovim version (nvim -v)

NVIM v0.8.0-dev-1088-gbd4eca114

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Linux Mint 20.2

### Terminal name/version

Kitty 0.15

### $TERM environment variable

xterm-kitty

### Installation

appimage

### How to reproduce the issue

1. `nvim --clean repro.lua`
2. Paste the following Lua code:
```lua
vim.cmd[[
set undofile
inoremap . .<C-g>u]]

local function command_preview(opts, _, _)
  vim.cmd("norm i" .. opts.args)
  return 0
end

vim.api.nvim_create_user_command("Test", function() end, { preview = command_preview, nargs = 1 })
```
3. `:so %`
4. Type `:Test a.a.a.a.a.`

### Expected behavior

The buffer text should not change because `0` is returned from the preview callback indicating that no preview should be shown. The undo file should not be corrupted.

### Actual behavior

Previewing the command has two unwanted side effects:
1. The buffer text is modified and `a.a.a.a.a.a.a.a.a.a.` is inserted (this change also can't be undone by pressing `u`).
2. Undoing the changes manually, then trying to source the file results in the following error: 
```
Error detected while processing /home/user/repro.lua:                                                        
E5113: Error while calling lua chunk: /home/user/test.lua:4: Vim(set):E825: Corrupted undo file (num_head): /home/j
onas/.local/state/nvim/undo/%home%user%repro.lua
stack traceback:
        [C]: in function 'cmd'
        /home/user/repro.lua:4: in main chunk
```

### PR DESCRIPTION
**Pull request by ateoi**
Problem:
The cmdpreview saved undo nodes on `cmdpreview_prepare` from `ex_getln.c` may become invalid (free) if the preview function makes undo operations, causing heap-use-after-free errors.

Solution:
Save the buffer undo list on `cmdpreview_prepare` and start a new empty one. On `cmdpreview_restore_state`, undo all the entries in the new undo list and restore the original one. With this approach, the preview function will be allowed to undo only its own changes.

fix #20036
fix #20248

### CODE REVIEW

**Comment by justinmk** [comment]
Thanks a lot. Hoping @famiu can review. Can you add a test?

**Comment by famiu** [comment]
Perhaps it would be a good idea to move the undo related save/restore to their own functions, to clean up `cmdpreview_prepare()` and `cmdpreview_restore_state()`

**Comment by famiu** [comment]
I'm not sure if not saving and restoring `b_u_save_nr_cur` and `b_u_save_nr_cur` could potentially cause problems, but that might be the case and is worth looking into

**Comment by ateoi** [comment]
> Perhaps it would be a good idea to move the undo related save/restore to their own functions that get called within the cmdpreview prepare/restore functions, to clean up `cmdpreview_prepare()` and `cmdpreview_restore_state()`

> I'm not sure if not saving and restoring `b_u_save_nr_cur` and `b_u_save_nr_last` could potentially cause problems, but that might be the case and is worth looking into

I've unified the undo state restoration from cmd preview callback and `undofile` to remove duplicate code. As both `b_u_save_nr_cur` and `b_u_save_nr_last` were restored by `undofile`, I've included them in the undo state save/restore functions.

**Comment by ateoi** [comment]
> I'm not sure if not saving and restoring `b_u_save_nr_cur` and `b_u_save_nr_last` could potentially cause problems, but that might be the case and is worth looking into

I guess not saving these fields could be a problem if the preview callback function make calls to `vim.cmd.write()`, but that can cause other weird behaviors. For example, the user command below will write a modified version of the buffer and this won't be undone when it leaves the preview callback.

```lua
vim.api.nvim_create_user_command('Foo', function() end, {
  nargs = '?',
  preview = function()
    vim.cmd('norm ifoo')
    vim.cmd.write()
  end,
})
```

File: src/nvim/undo.c [hunk 1]
```diff
@@ -1615,20 +1615,26 @@ void u_read_undo(char *name, const uint8_t *hash, const char *orig_name FUNC_ATT
...
+    .save_b_u_line_ptr = line_ptr,
+    .save_b_u_line_lnum = line_lnum,
+    .save_b_u_line_colnr = line_colnr,
+    .save_b_p_ul = curbuf->b_p_ul,
+    .save_b_changed = curbuf->b_changed,
+    .save_changedtick = buf_get_changedtick(curbuf),
+    .save_b_u_save_nr_cur = last_save_nr,
+    .save_b_u_save_nr_last = last_save_nr,
+  };
+  u_restore_state(&undostate);
```
Comment (zeertzjq) [inline src/nvim/undo.c:1637]: What is the point of this change? It seems to make code more confusing.

Comment (ateoi) [reply src/nvim/undo.c]: The goal was to add a generic interface to save/restore the undo state.

Comment (justinmk) [reply src/nvim/undo.c]: The `save_*` prefix on each field seems redundant?

Comment (ateoi) [reply src/nvim/undo.c]: > The `save_*` prefix on each field seems redundant?

Agree. And if we remove the prefix, we could use it to replace the `b_u_*`` loose fields from `file_buffer`.

https://github.com/neovim/neovim/blob/ccf328172bac2b02f9bd19fa58e105958514a28a/src/nvim/buffer_defs.h#L568-L583

Comment (famiu) [reply src/nvim/undo.c]: > > The `save_*` prefix on each field seems redundant?
> 
> Agree. And if we remove the prefix, we could use it to replace the `b_u_*` loose fields from `file_buffer`.
> 
> https://github.com/neovim/neovim/blob/ccf328172bac2b02f9bd19fa58e105958514a28a/src/nvim/buffer_defs.h#L568-L583

I like that idea a lot. Do you plan on making a PR for it?

Comment (zeertzjq) [reply src/nvim/undo.c]: I also thought these `b_u_` fields can moved into a structs with slightly different names, and then `#define`s can be added for the original names, but I had some trouble giving a name to the struct and its fields.

**Pull request merged**